### PR TITLE
Added support for using openldap as a backend for identity driver, as…

### PR DESCRIPTION
**Pull request by MaheshIBM**
…signment will be still be sql driver, password policies for user passwords can be enabled, and password syntax check controlled.

### CODE REVIEW

File: envs/example/defaults.yml [hunk 1]
```diff
@@ -389,6 +389,29 @@ keystone:
...
+  suffix: "dc=example,dc=org"
+  users_dn: "ou=users,ou=openstack,dc=example,dc=org"
+  groups_dn: "ou=groups,ou=openstack,dc=example,dc=org"
+
+  # keystone user settings, this is used by the keystone ldap driver to
+  # connect to the ldap server, it needs to have read and write access
+  # to the user and group sub trees as set by openldap.users_dn
+  # and openldap.groups_dn respectively
+  openldap_user_name: keystone_user
+  openldap_user_password: aq1sw2de
```
Comment (paulczar) [inline envs/example/defaults.yml]: these passwords should probably be stored in the `secrets` namespace and imported in ?

File: roles/openldap-remove-password-policy/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+- name: Run remove policy for service users
+  command: "{{openldap_password_policy.config_dir}}password_policy_util.sh -r {{ item.name }}"
```
Comment (paulczar) [inline roles/openldap-remove-password-policy/tasks/main.yml]: for readability the joining slash for paths like this should be in the string,  not in the variable.    e.g. `{{openldap_password_policy.config_dir}}/password_policy_util.sh`

File: roles/openldap/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+- include: password_policy.yml
+  when: openldap.password_policy_enabled|bool
+
+- include: sync_provider.yml
+  when: groups['ldap'] | length > 1
+
+
+- file: path="{{openldap.openldap_config_dir}}" state=directory mode=0755
+- name: Create password policy utility script
+  template: src=password_policy_util.sh
```
Comment (paulczar) [inline roles/openldap/tasks/main.yml]: this contains the ldap root password ...  mode should not allow reading by nonroot or executable by nonroot ... so should probably be `0700`

**Comment by j2sol** [comment]
This commit message needs some re-formatting

http://chris.beams.io/posts/git-commit/

File: roles/openldap/tasks/tmp_man/remove_password_policy.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+#- hosts: 192.168.0.106
```
Comment (paulczar) [inline roles/openldap/tasks/tmp_man/remove_password_policy.yml]: what is `tmp_man`  ??

File: roles/openldap/templates/etc/init/slapd.conf [hunk 5]
```diff
@@ -0,0 +1,27 @@
+description "slapd - Standalone LDAP daemon"
```
Comment (paulczar) [inline roles/openldap/templates/etc/init/slapd.conf]: can we use the `upstart_service` module for this ?

File: roles/openldap/templates/etc/openldap/ldif_config_files/empty_password_policy.ldif [hunk 6]
```diff
@@ -0,0 +1,15 @@
+dn: cn=nopolicy,ou=policies,ou=openstack,dc=example,dc=org
```
Comment (paulczar) [inline roles/openldap/templates/etc/openldap/ldif_config_files/empty_password_policy.ldif]: shound't dn be templated from vars ?

File: roles/openldap/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+- name: template config ldif files
+  template: src={{ item }}
+            dest={{ openldap.config_dir }}
+  with_fileglob: ../templates/{{ openldap.config_dir }}/*
+
+- name: ensure admin/suffix is created
+  shell: ldapmodify -Y EXTERNAL -H ldapi:///
+    -f {{ openldap.config_dir }}/set_admin.ldif
+  register: set_admin
+  failed_when: set_admin.rc != 0 and set_admin.rc != 20
```
Comment (j2sol) [inline roles/openldap/tasks/main.yml]: `failed_when: set_admin.rc not in [0, 20]`

File: roles/openldap/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+
+
+- name: ensure base domain is created
+  shell: ldapadd -D {{ openldap.root_dn }} -w {{ openldap.root_password }}
+    -f {{ openldap.config_dir }}/base_domain.ldif
+  register: base_domain
+  failed_when: base_domain.rc != 0 and base_domain.rc != 68
```
Comment (j2sol) [inline roles/openldap/tasks/main.yml]: `base_domain.rc not in [0, 68]`

File: roles/openldap/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+- name: ensure keystone_user is created
+  shell: ldapadd -D {{ openldap.root_dn }} -w {{ openldap.root_password }}
+    -f {{ openldap.config_dir }}/keystone_user.ldif
+  register: keystone_user
+  failed_when: keystone_user.rc != 0 and keystone_user.rc != 68
```
Comment (j2sol) [inline roles/openldap/tasks/main.yml]: `keystone_user.rc not in [0, 68]`

File: roles/openldap/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+- name: create upstart config for slapd
+  template: src=etc/init/slapd.conf
+            dest=/etc/init/slapd.conf
+  register: slapd_upstart
+
+  #slapd is setup not to use upstart
+  #so this step is needed
+- name: stop using init.d for slapd
+  shell: /etc/init.d/slapd stop && initctl reload-configuration slapd && service slapd restart
+  when: slapd_upstart.changed or slapd_envs.changed
```
Comment (j2sol) [inline roles/openldap/tasks/main.yml]: Will the `slapd stop` call result in an error if the service isn't already running?

File: roles/openldap/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+- name: delete the policy if policy is changed from enabled to disabled.
+  command: ldapdelete -x -D {{openldap.root_dn }} -w {{ openldap.root_password }} {{ openldap.password_policy_dn }}
+  register: command_result
+  when: openldap.password_policy_enabled|bool == False
+  failed_when: command_result.rc != 32 and command_result.rc!=0
```
Comment (j2sol) [inline roles/openldap/tasks/main.yml]: `command_result.rc not in [0, 32]`

Comment (j2sol) [inline roles/openldap/tasks/main.yml]: All of these where you're checking against multiple return codes could use comments about what the return code you're calling out is for. It'll be helpful down the road.

File: roles/openldap/tasks/password_policy.yml [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+  ignore_errors: yes
+  changed_when: False
+
+- name: add password policy schema
+  shell: ldapadd -Y EXTERNAL -H ldapi:/// -f  /etc/ldap/schema/ppolicy.ldif
+  when: ppolicy_out.stdout.count("ppolicy,") < 1
+
+- name: create policy container
+  shell: ldapadd -x -D {{openldap.root_dn }} -w {{ openldap.root_password }} 
+    -f {{ openldap.config_dir }}/ppolicy_container.ldif
```
Comment (j2sol) [inline roles/openldap/tasks/password_policy.yml]: This looks like it needs to be whitespaced over a bit, same with other shell tasks in this pr. For multiline the second line should start at the same level as the previous line

```
  shell: ldapadd -x -D {{openldap.root_dn }} -w {{ openldap.root_password }} 
         -f {{ openldap.config_dir }}/ppolicy_container.ldif
```

File: roles/openldap/tasks/password_policy.yml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  register: policy_command
+  failed_when: policy_command.rc != 0 and policy_command.rc != 68
```
Comment (j2sol) [inline roles/openldap/tasks/password_policy.yml]: `policy_command.rc not in [0, 68]`

File: roles/openldap/tasks/password_policy.yml [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+- name: add password policy module
+  shell: ldapadd -Y EXTERNAL -H ldapi:/// -f {{ openldap.config_dir }}/ppolicy_module.ldif
+  when: pm_out.stdout.count(",ou={{ openldap.password_policy_ou }}") < 1
+
+- name: add default policy
+  shell: ldapadd -x -D {{openldap.root_dn }} -w {{ openldap.root_password }} 
+    -f {{ openldap.config_dir }}/password_policy.ldif
+  register: default_policy
+  failed_when: default_policy.rc != 0 and default_policy.rc != 68
```
Comment (j2sol) [inline roles/openldap/tasks/password_policy.yml]: `default_policy.rc not in [0, 68]`

File: roles/openldap/tasks/password_policy.yml [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+- name: add empty policy 
+  shell: ldapadd -x -D {{openldap.root_dn }} -w {{ openldap.root_password }}
+    -f {{ openldap.config_dir }}/empty_password_policy.ldif
+  register: empty_policy
+  failed_when: empty_policy.rc != 0 and empty_policy.rc != 68
```
Comment (j2sol) [inline roles/openldap/tasks/password_policy.yml]: `empty_policy not in [0, 68]`

File: roles/openldap/tasks/password_policy.yml [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+- name: install checker from the package
+  shell: dpkg -i /tmp/openldap-password-checker-1.0-1.deb
+
+#END TODO Install from repo its uploaded there
+
+- name: add check_password ldif into ldap
+  command: /usr/bin/ldapmodify -x -D {{ openldap.root_dn }} 
+    -w {{ openldap.root_password }} -f {{ openldap.config_dir }}/add_passwd_config.ldif
+  register: check_pass
+  failed_when: check_pass.rc !=0 and check_pass.rc != 20
```
Comment (j2sol) [inline roles/openldap/tasks/password_policy.yml]: `check_pass.rc not in [0, 20]`

File: roles/openldap/tasks/password_policy.yml [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+
+- name: create check_password config directory
+  file: path=/etc/openldap 
+        state=directory
+
+- name: create check_password config
+  template: src=check_password.conf
```
Comment (j2sol) [inline roles/openldap/tasks/password_policy.yml]: template sources should match template destination. The source for this should be `etc/openldap/check_password.conf` and it should live in that relative path within the role's template directory.

File: roles/openldap/tasks/tmp_man/remove_password_policy.yml [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+#  tasks:
+#    - name: read a file
+     # command: cat /root/development/mahesh/ursula/roles/openldap/tasks/tmp_man/sample.txt
+#      command: cat /root/man/sample.txt
+#      register: command_result
+#      with_lines: "{{text}}"
+#    - debug: msg={{ command_result.stdout  }}
+
+- hosts: 192.168.0.106
```
Comment (j2sol) [inline roles/openldap/tasks/tmp_man/remove_password_policy.yml]: What is this?

File: roles/openldap/templates/password_policy_util.sh [hunk 19]
```diff
@@ -0,0 +1,176 @@
+#!/bin/bash
```
Comment (j2sol) [inline roles/openldap/templates/password_policy_util.sh]: Is this file intended to be used outside of Ansible? If not, we shouldn't install it on the host, and instead turn it into an Ansible module we can use to accomplish the tasks.

**Comment by j2sol** [comment]
Various comments within.

We need to resolve the check_password library before merging this, as we've discussed.

**Comment by j2sol** [comment]
Holding off on this work as we're investigating Wafflehaus as being a lighter way of enforcing (some) password requirements for (some) users.

File: roles/openldap/templates/password_policy_util.sh (hunk 19 above)
Comment (MaheshIBM) [inline roles/openldap/templates/password_policy_util.sh]: Yes its intended to be used outside ansible as a helper script to unlock users/revoke password policy  of users etc.

**Comment by paulczar** [comment]
closing all PRs with no recent activity.
